From 9e871fce2cd86c1e7ee7db0b7f658b63c6b70c84 Mon Sep 17 00:00:00 2001
From: Phil Sutter <psutter@redhat.com>
Date: Sat, 2 Jul 2016 12:45:04 +0200
Subject: [PATCH] iplink_geneve: add ttl configuration at link creation

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1339178
Upstream Status: iproute2.git commit f4c05c2e99538
Conflicts: Context changes due to already updated kernel headers.

commit f4c05c2e9953810c42e37ae21f6970c1f024fa46
Author: John W. Linville <linville@tuxdriver.com>
Date:   Mon Jun 15 14:37:15 2015 -0400

    iplink_geneve: add ttl configuration at link creation

    Signed-off-by: John W. Linville <linville@tuxdriver.com>
---
 ip/iplink_geneve.c    | 23 ++++++++++++++++++++++-
 man/man8/ip-link.8.in |  7 +++++++
 2 files changed, 29 insertions(+), 1 deletion(-)

diff --git a/ip/iplink_geneve.c b/ip/iplink_geneve.c
index 74703e1..dff08a6 100644
--- a/ip/iplink_geneve.c
+++ b/ip/iplink_geneve.c
@@ -17,9 +17,11 @@
 static void print_explain(FILE *f)
 {
 	fprintf(f, "Usage: ... geneve id VNI remote ADDR\n");
+	fprintf(f, "                 [ ttl TTL ]\n");
 	fprintf(f, "\n");
 	fprintf(f, "Where: VNI  := 0-16777215\n");
 	fprintf(f, "       ADDR := IP_ADDRESS\n");
+	fprintf(f, "       TTL  := { 1..255 | inherit }\n");
 }
 
 static void explain(void)
@@ -34,7 +36,7 @@ static int geneve_parse_opt(struct link_util *lu, int argc, char **argv,
 	int vni_set = 0;
 	__u32 daddr = 0;
 	struct in6_addr daddr6 = IN6ADDR_ANY_INIT;
-
+	__u8 ttl = 0;
 
 	while (argc > 0) {
 		if (!matches(*argv, "id") ||
@@ -52,6 +54,18 @@ static int geneve_parse_opt(struct link_util *lu, int argc, char **argv,
 			}
 			if (IN_MULTICAST(ntohl(daddr)))
 				invarg("invalid remote address", *argv);
+		} else if (!matches(*argv, "ttl") ||
+			   !matches(*argv, "hoplimit")) {
+			unsigned uval;
+
+			NEXT_ARG();
+			if (strcmp(*argv, "inherit") != 0) {
+				if (get_unsigned(&uval, *argv, 0))
+					invarg("invalid TTL", *argv);
+				if (uval > 255)
+					invarg("TTL must be <= 255", *argv);
+				ttl = uval;
+			}
 		} else if (matches(*argv, "help") == 0) {
 			explain();
 			return -1;
@@ -80,6 +94,7 @@ static int geneve_parse_opt(struct link_util *lu, int argc, char **argv,
 	addattr32(n, 1024, IFLA_GENEVE_ID, vni);
 	if (daddr)
 		addattr_l(n, 1024, IFLA_GENEVE_REMOTE, &daddr, 4);
+	addattr8(n, 1024, IFLA_GENEVE_TTL, ttl);
 
 	return 0;
 }
@@ -105,6 +120,12 @@ static void geneve_print_opt(struct link_util *lu, FILE *f, struct rtattr *tb[])
 			fprintf(f, "remote %s ",
 				format_host(AF_INET, 4, &addr, s1, sizeof(s1)));
 	}
+
+	if (tb[IFLA_GENEVE_TTL]) {
+		__u8 ttl = rta_getattr_u8(tb[IFLA_GENEVE_TTL]);
+		if (ttl)
+			fprintf(f, "ttl %d ", ttl);
+	}
 }
 
 static void geneve_print_help(struct link_util *lu, int argc, char **argv,
diff --git a/man/man8/ip-link.8.in b/man/man8/ip-link.8.in
index 21503ff..284ee13 100644
--- a/man/man8/ip-link.8.in
+++ b/man/man8/ip-link.8.in
@@ -670,6 +670,9 @@ the following additional arguments are supported:
 
 .BI "ip link add " DEVICE
 .BI type " geneve " id " ID " remote " IPADDR"
+.R " [ "
+.BI ttl " TTL "
+.R " ]"
 
 .in +8
 .sp
@@ -680,6 +683,10 @@ the following additional arguments are supported:
 .BI remote " IPADDR"
 - specifies the unicast destination IP address to use in outgoing packets.
 
+.sp
+.BI ttl " TTL"
+- specifies the TTL value to use in outgoing packets.
+
 .in -8
 
 .TP
-- 
1.8.3.1

